test/test_torch.py

```diff
@@ -17485,27 +17485,75 @@ def test_rdiv(self, device, dtype):
+                    mod = mod.to(dtype)
+                mod = mod.cpu().numpy()
+            else:
+                if dtype in torch.testing.get_all_int_dtypes():
+                    mod = int(mod)
+            np_x = x.cpu().numpy()
+            exp = np.fmod(np_x, mod)
+            exp = torch.from_numpy(exp)
+            res = res.to(exp.dtype)
+            self.assertEqual(res, exp)
+
+        x = make_tensor((10, 10), device=device, dtype=dtype, low=-9, high=9)
+        # non-contiguours
+        xt = x.t()
+        # Exclude 0
+        mod = make_tensor((10, 10), device=device,
+                          dtype=torch.float if dtype in torch.testing.get_all_int_dtypes() else dtype,
+                          low=1, high=9)
+        zero = torch.zeros_like(x)
+
+        # Integer Scalar
+        _reference_implementation(dtype, x, 3)
+        _reference_implementation(dtype, xt, 3)
+        # Float Scalar
+        _reference_implementation(dtype, x, 2.3)
+        _reference_implementation(dtype, xt, 2.3)
 
-        zero = torch.zeros_like(m1)
         if dtype in torch.testing.get_all_int_dtypes():
-            with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
-                m1.fmod(0)
-            with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
-                m1.fmod(zero)
+            # Float Tensor
+            with self.assertRaisesRegex(RuntimeError, "result type (Half|Float|Double) "
+                                                      "can't be cast to the desired "
+                                                      "output type (Byte|Char|Short|Int|Long)"):
+                x.fmod(mod)
+            # fmod to 0
+            if device == 'cpu':
+                # fmod to 0
+                with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
+                    x.fmod(0)
+                with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
+                    x.fmod(zero)
+            else:
+                if dtype == torch.int64:
+                    self.assertEqual(x.fmod(0) == 4294967295, x >= 0)
+                    self.assertEqual(x.fmod(0) == -1, x < 0)
+                    self.assertEqual(x.fmod(zero) == 4294967295, x >= 0)
+                    self.assertEqual(x.fmod(zero) == -1, x < 0)
+                else:
+                    value = 255 if dtype == torch.uint8 else -1
+                    self.assertTrue(torch.all(x.fmod(0) == value))
+                    self.assertTrue(torch.all(x.fmod(zero) == value))
         else:
-            self.assertTrue(torch.all(m1.fmod(0).isnan()))
-            self.assertTrue(torch.all(m1.fmod(zero).isnan()))
+            # Float Tensor
+            _reference_implementation(dtype, x, mod)
+            _reference_implementation(dtype, xt, mod)
+            # fmod to 0
+            self.assertTrue(torch.all(x.fmod(0).isnan()))
+            self.assertTrue(torch.all(x.fmod(zero).isnan()))
```

Comment:
nit: All of these nested if-then-else can be hard to read. How about splitting these off into different tests?

e.g.

```
@dtypes(floating_point)
def test_fmod_by_zero_float(self, dtype, device):
    # check fmod to zero is nan

@dtypes(integral)
def test_fmod_by_zero_integral(self, dtype, device):
    # check fmod to zero is Zero div error on CPU and weird number on GPU
```